opendevin/memory/condenser.py

```diff
@@ -1,26 +1,199 @@
+from typing import Callable
+
 from opendevin.core.logger import opendevin_logger as logger
+from opendevin.core.utils import json
 from opendevin.llm.llm import LLM
 
+MAX_TOKEN_COUNT_PADDING = 1024
+
 
 class MemoryCondenser:
-    def condense(self, summarize_prompt: str, llm: LLM):
+    """
+    Condenses the prompt with a call to the LLM.
+    """
+
+    def __init__(
+        self,
+        action_prompt: Callable[..., str],
+        summarize_prompt: Callable[[list[dict], list[dict]], str],
+    ):
+        """
+        Initialize the MemoryCondenser with the action and summarize prompts.
+
+        action_prompt is a callable that returns the prompt that is about to be sent to the LLM.
+        The prompt callable will be called with default events and recent events as arguments.
+        summarize_prompt, which is optional, is a callable that returns a specific prompt to tell the LLM to summarize the recent events.
+        The prompt callable will be called with default events and recent events as arguments.
+
+        Parameters:
+        - action_prompt (Callable): The function to generate an action prompt. The function should accept core events and recent events as arguments.
+        - summarize_prompt (Callable): The function to generate a summarize prompt. The function should accept core events and recent events as arguments.
+        """
+        self.action_prompt = action_prompt
+        self.summarize_prompt = summarize_prompt
+
+    def condense(
+        self,
+        llm: LLM,
+        default_events: list[dict],
+        recent_events: list[dict],
+        background_commands: list | None = None,
+    ) -> tuple[list[dict], bool]:
         """
-        Attempts to condense the monologue by using the llm
+        Attempts to condense the recent events of the monologue by using the llm, if necessary. Returns the condensed recent events if successful, or False if not.
+
+        It includes default events in the prompt for context, but does not alter them.
+        Condenses the monologue using a summary prompt.
+        Checks if the action_prompt (including events) needs condensation based on token count, and doesn't attempt condensing if not.
+        Returns unmodified list of recent events if it is already short enough.
 
         Parameters:
-        - llm (LLM): llm to be used for summarization
+        - llm (LLM): LLM to be used for summarization.
+        - default_events (list[dict]): List of default events that should remain unchanged.
+        - recent_events (list[dict]): List of recent events that may be condensed.
+        - background_commands (list): List of background commands to be included in the prompt.
 
-        Raises:
-        - Exception: the same exception as it got from the llm or processing the response
+        Returns:
+        - list[dict] | bool: The condensed recent events if successful, unmodified list if unnecessary, or False if condensation failed.
```

Comment:
Seems this docstring is outdated? Return value is of `tuple[list[dict], bool]` type

Reply:
We probably shouldn't include types in the docstring, as they're already included in the type annotations and duplication is a source of failure. I'd suggest just deleting types from docstrings.

Reply:
Cleaned them up!